### PR TITLE
drivers: ethernet: sam_gmac: simplify ETH_SAM_GMAC_QUEUES Kconfig

### PR DESCRIPTION
**Pull request by galak**
Make the max range for ETH_SAM_GMAC_QUEUES Kconfig symbol a fixed value
rather than coming from devicetree.  We than add an error check to
verify the value against the device tree 'num-queues' property in the
driver itself.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

### CODE REVIEW

**Comment by galak** [comment]
I wonder if this should actually be:

`#if CONFIG_ETH_SAM_GMAC_QUEUES >= DT_INST_PROP(0, num_queues)`

If I set it to 3 on sam_e70_xplained I get:
```
/home/galak/git/zephyr/drivers/ethernet/eth_sam_gmac.c: In function 'eth0_iface_init':
/home/galak/git/zephyr/drivers/ethernet/eth_sam_gmac.c:287:20: warning: array subscript 2 is above array bounds of 'volatile uint32_t[2]' {aka 'volatile unsigned int[2]'} [-Warray-bounds]
  287 |  gmac->GMAC_RBQBAPQ[queue->que_idx - 1] = (u32_t)rx_desc_list->buf;
      |  ~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
/home/galak/git/zephyr/drivers/ethernet/eth_sam_gmac.c:289:20: warning: array subscript 2 is above array bounds of 'volatile uint32_t[2]' {aka 'volatile unsigned int[2]'} [-Warray-bounds]
  289 |  gmac->GMAC_TBQBAPQ[queue->que_idx - 1] = (u32_t)tx_desc_list->buf;
```

File: drivers/ethernet/Kconfig.sam_gmac [hunk 1]
```diff
@@ -15,7 +15,7 @@ if ETH_SAM_GMAC
 config ETH_SAM_GMAC_QUEUES
 	int "Number of active hardware TX and RX queues"
 	default 1
-	range 1 $(dt_node_int_prop_int,/soc/ethernet@40050088,num-queues)
+	range 1 6
```
Comment (stephanosio) [inline drivers/ethernet/Kconfig.sam_gmac:18]: I think it would be better to just remove `range` here since this can be confusing/misleading, especially when using menuconfig.

The value is going to be checked in the driver code anyway.

File: drivers/ethernet/eth_sam_gmac.c [hunk 2]
```diff
@@ -49,6 +49,10 @@ LOG_MODULE_REGISTER(LOG_MODULE_NAME);
 /*
  * Verify Kconfig configuration
  */
+#if CONFIG_ETH_SAM_GMAC_QUEUES > DT_INST_PROP(0, num_queues)
+#error Number of active hardware TX and RX queues is invalid
+#endif
```
Comment (stephanosio) [inline drivers/ethernet/eth_sam_gmac.c:54]: I wonder if using `BUILD_ASSERT` would be better.

One for "Number of active hardware TX and RX queues cannot be 0"
Another for "Number of active hardware TX and RX queues is invalid" or "... exceeds the maximum queue count supported in hardware"

**Comment by stephanosio** [comment]
> `#if CONFIG_ETH_SAM_GMAC_QUEUES >= DT_INST_PROP(0, num_queues)`
> 
> If I set it to 3 on sam_e70_xplained I get:

Strange ...

When `CONFIG_ETH_SAM_GMAC_QUEUES=3`:

https://github.com/zephyrproject-rtos/zephyr/blob/071a986bb676dc10252f92052ba056cc47717195/drivers/ethernet/eth_sam_gmac_priv.h#L41-L42

* `GMAC_ACTIVE_QUEUE_NUM` = 3
* `GMAC_ACTIVE_PRIORITY_QUEUE_NUM` = 2

Since DT `num-queues` = `GMAC_QUEUE_NUM` = 3, `queue_init` is called for `que_idx` = 0 .. 2:

https://github.com/zephyrproject-rtos/zephyr/blob/071a986bb676dc10252f92052ba056cc47717195/drivers/ethernet/eth_sam_gmac.c#L1908-L1909

Since `GMAC_ACTIVE_PRIORITY_QUEUE_NUM` = 2, `priority_queue_init` will be called for `que_idx` = 1 .. 2:

https://github.com/zephyrproject-rtos/zephyr/blob/071a986bb676dc10252f92052ba056cc47717195/drivers/ethernet/eth_sam_gmac.c#L296-L305

In `priority_queue_init`, we set `queue_index = queue->que_idx - 1` and use it to access the priority queue registers indexes 0 .. 1:

https://github.com/zephyrproject-rtos/zephyr/blob/071a986bb676dc10252f92052ba056cc47717195/drivers/ethernet/eth_sam_gmac.c#L225-L226

https://github.com/zephyrproject-rtos/zephyr/blob/071a986bb676dc10252f92052ba056cc47717195/drivers/ethernet/eth_sam_gmac.c#L247-L250

In other words, we should not be accessing the index 2, not sure why GCC is complaining about it.

Feel free to double check.

**Comment by galak** [comment]
@stephanosio can I have you take over this PR and make whatever changes you see fit.

**Comment by stephanosio** [comment]
> @stephanosio can I have you take over this PR and make whatever changes you see fit.

alright, will do.

**Comment by stephanosio** [comment]
Closing this PR as the issue has been addressed in #23031 and #23300.
